HW7 Architecture

Mishok7889 opened the pull request.

Commit:
Refactor project.

Commit:
Fix mock generation.

Commit:
Fix test fails on github action.

Commit:
Fix e2e tests.

Commit:
Add architecture tests.

Commit:
Improve ports and adapters architecture.

- Add proper error handling for ports.NotFoundError in getWeatherFromProvider
- Convert ports.NotFoundError to shared.NewNotFoundError to maintain proper error flow
- Fix e2e test failure where weather API returned 500 instead of 404 for invalid cities
- Ensure correct HTTP status code mapping through architectural layers

Previously, ports.NotFoundError from external weather providers was not being
properly handled and fell through to external service error handling, resulting
in 500 status codes instead of the expected 404 for non-existent cities.

## Major Restructuring
- Restructure application ports into bounded contexts (Weather, Subscription, Notification, Infrastructure)
- Add service adapters for cross-context communication (WeatherServiceAdapter, SubscriptionServiceAdapter)
- Replace pkg/errors with layered error types (shared.DomainError, infrastructure.InfrastructureError, api.APIError)
- Implement proper separation between domain, ports, and adapters layers
- Refactor dependency injection to support bounded context isolation

## Architecture Implementation
- Add service ports for cross-context communication
- Implement proper port interfaces and data transfer objects
- Remove direct domain type sharing between bounded contexts
- Replace direct slog usage with logger ports throughout application
- Move infrastructure concerns to dedicated infrastructure layer

## Structural Changes
- Complete API layer refactoring (handlers, error handling, server setup)
- External adapters restructuring (weather providers, cache, email, database)
- Core domain restructuring (weather, subscription, notification use cases)
- Application bootstrapping and dependency injection overhaul
- Configuration system standardization

Commit:
Add diagram.

Commit:
Add architecture diagram.

YuliiaAndreieva commented on internal/adapters/api/error_handler_test.go:
looks like in in this tests file you have tests that difference only params in NewRequest and http responce, maybe you can reorganize it in table-driven tests
```go
tests := []struct {
	route           string
	expectedStatus  int
	expectedMessage string
}{
	{"/test/validation", http.StatusBadRequest, "validation failed"},
	{"/test/not-found", http.StatusNotFound, "resource not found"},
	// ...
}

YuliiaAndreieva commented on internal/adapters/api/subscription_handler.go:
```diff
@@ -0,0 +1,106 @@
+func (s *HTTPServerAdapter) confirmSubscription(c *gin.Context) {
+	token := c.Param("token")
+	if token == "" {
+		s.handleError(c, NewValidationError("token parameter is required"))
```
duplicate `token parameter is required` , I think it's better to have params validation is some other place

YuliiaAndreieva commented on internal/adapters/api/subscription_handler_test.go:
```diff
@@ -0,0 +1,433 @@
+}
+
+func TestSubscriptionHandler_Subscribe_InvalidEmail(t *testing.T) {
+	router, _, _, _ := setupSubscriptionTestRouter(t)
```
same as last pr

YuliiaAndreieva commented on internal/adapters/api/subscription_handler_test.go:
```diff
@@ -0,0 +1,433 @@
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
```
![cat](https://github.com/user-attachments/assets/8ebbc306-871f-42bd-9f17-b673971eec5b)

YuliiaAndreieva commented on internal/adapters/database/subscription_repo.go:
```diff
@@ -0,0 +1,215 @@
+}
+
+// FindConfirmedByFrequency retrieves all confirmed subscriptions for a specific frequency
+func (r *SubscriptionRepositoryAdapter) FindConfirmedByFrequency(ctx context.Context, frequency string) ([]*ports.SubscriptionData, error) {
```
looks like you have a typo, in `FindActiveByFrequency` you have `GetConfirmedByFrequency` call also, but if it's not a typo maybe it's better to have one method for it because it will a lot  of methods later like FindActiveByEmail and so on..

Mishok7889 replied:
Thank you. I upgraded the solution using a filter pattern.

YuliiaAndreieva commented on internal/adapters/database/subscription_repo.go:
duplicate error string like subscription cannot be nil

YuliiaAndreieva commented on internal/adapters/database/token_repo.go:
```diff
@@ -0,0 +1,198 @@
+	}
+
+	token := &ports.TokenData{
+		Value:          uuid.New().String(),
```
I think repo should not know how token is generated

YuliiaAndreieva commented on internal/adapters/database/token_repo.go:
```diff
@@ -0,0 +1,198 @@
+}
+
+// CreateUnsubscribeToken creates a new unsubscribe token
+func (r *TokenRepositoryAdapter) CreateUnsubscribeToken(ctx context.Context, subscriptionID uint, expiresIn time.Duration) (*ports.TokenData, error) {
```
you already have `CreateConfirmationToken`, I suppose they differs only the token type (which better to have constant)

YuliiaAndreieva commented on internal/adapters/external/email_provider.go:
```diff
@@ -0,0 +1,144 @@
+}
+
+// SendEmail sends an email using SMTP with flexible authentication and TLS
+func (p *SMTPEmailProviderAdapter) SendEmail(ctx context.Context, params ports.EmailParams) error {
```
maybe divide it in some methods

YuliiaAndreieva commented on internal/adapters/external/email_provider.go:
```diff
@@ -0,0 +1,144 @@
+		return infrastructure.NewValidationError("email body cannot be empty")
+	}
+
+	from := fmt.Sprintf("%s <%s>", p.fromName, p.fromAddr)
```
I think it's closer to the building email logic, not for sending

YuliiaAndreieva commented on internal/adapters/external/email_provider.go:
```diff
@@ -0,0 +1,144 @@
+}
+
+// ValidateConfiguration validates the email provider configuration
+func (p *SMTPEmailProviderAdapter) ValidateConfiguration() error {
```
where do u use it besides test ?

anastasiia-tokareva-hw replied:
Looks like it should be used in `NewSMTPEmailProviderAdapter`

YuliiaAndreieva commented on internal/adapters/external/email_provider_test.go:
```diff
@@ -0,0 +1,234 @@
+	"weatherapi.app/internal/ports"
+)
+
+func TestSMTPEmailProviderAdapter_ValidateConfiguration(t *testing.T) {
```
I think you can try to create one valid config and modified the field that requires validation error to reduce duplication, what I mean:
```go
validConfig := EmailProviderConfig{
		Host:     "smtp.example.com",
		Port:     587,
		Username: "user",
		Password: "pass",
		FromName: "App",
		FromAddr: "app@company.com",
	}

	tests := []struct {
		name        string
		modify      func(cfg EmailProviderConfig) EmailProviderConfig
		expectError bool
	}{
		{
			name: "Missing Host",
			modify: func(cfg EmailProviderConfig) EmailProviderConfig {
				cfg.Host = ""
				return cfg
			},
			expectError: true,
		},
```
That's idea from [here](https://www.willem.dev/articles/large-structs-in-table-tests/#:~:text=If%20we%20apply%20this%20to%20our%20test%20cases%20it%20looks%20like%20this%3A)

YuliiaAndreieva commented on internal/adapters/external/memory_cache_provider.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+func (c *MemoryCacheProvider) Set(ctx context.Context, key string, value []byte, ttl time.Duration) error {
+	if key == "" {
```
add method for validation

YuliiaAndreieva commented on internal/adapters/external/memory_cache_provider.go:
```diff
@@ -0,0 +1,152 @@
+
+func (c *MemoryCacheProvider) Exists(ctx context.Context, key string) (bool, error) {
+	if key == "" {
+		return false, infrastructure.NewValidationError("cache key cannot be empty")
```
again duplicate error string

YuliiaAndreieva commented on internal/adapters/external/redis_cache_provider.go:
```diff
@@ -0,0 +1,189 @@
+)
+
+// RedisCacheProviderAdapter implements CacheProvider port using Redis
+type RedisCacheProviderAdapter struct {
```
im not sure that you have redis client with metrics already is a good idea, would not it better to have one redis and then adapter that use redis and has metrics

YuliiaAndreieva commented on internal/adapters/external/redis_cache_provider_test.go:
```diff
@@ -0,0 +1,496 @@
+				if tt.checkError != nil {
+					tt.checkError(t, err)
+				}
+			} else {
```
better to have positive and negative tests separately

YuliiaAndreieva commented on internal/adapters/external/weather_cache_adapter_test.go:
```diff
@@ -0,0 +1,195 @@
+			require.NoError(t, err)
+			assert.Equal(t, weatherData, retrievedData)
+
+			// Test metrics through generic cache provider
```
i would separate this test because it hard to read

YuliiaAndreieva commented on internal/adapters/external/weather_providers.go:
```diff
@@ -0,0 +1,99 @@
+	return &WeatherAPIProviderAdapter{
+		apiKey:  params.APIKey,
+		baseURL: params.BaseURL,
+		client:  &http.Client{Timeout: 10 * time.Second},
```
magic number

anastasiia-tokareva-hw replied:
`const timeout = 10 * time.Second`

YuliiaAndreieva commented on internal/adapters/api/error_handler.go:
```diff
@@ -0,0 +1,133 @@
+	var statusCode int
+	var message string
+
+	// Handle API-specific errors
```
maybe separate in three methods like mapAPIErr mapPortErr

YuliiaAndreieva commented on internal/adapters/services/subscription_service_adapter.go:
```diff
@@ -0,0 +1,88 @@
+}
+
+// GetActiveSubscriptions returns all active subscriptions for a given frequency
+func (a *SubscriptionServiceAdapter) GetActiveSubscriptions(ctx context.Context, frequency string) ([]*ports.SubscriptionServiceData, error) {
```
same as I said before about repo methods

YuliiaAndreieva commented on internal/adapters/services/weather_service_adapter.go:
```diff
@@ -0,0 +1,42 @@
+
+// WeatherServiceAdapter adapts the weather use case to the WeatherService port
+type WeatherServiceAdapter struct {
+	weatherUseCase *weather.UseCase
```
Idk if it's ok to rely on instance, I created the interface for my usecase

anastasiia-tokareva-hw replied:
Yes, please use an interface

YuliiaAndreieva commented on internal/app/application.go:
```diff
@@ -0,0 +1,321 @@
+	return nil
+}
+
+func (a *Application) initializeAdapters() error {
```
I like that you divided initialize logic into different methods

YuliiaAndreieva commented on internal/config/config.go:
```diff
@@ -0,0 +1,323 @@
+	maxRedisDB         = 15
+	maxCacheTTLMinutes = 1440
+	maxDailyInterval   = 10080
+	maxPortNumber      = 65535
```
great job!

YuliiaAndreieva commented on internal/core/notification/usecase.go:
```diff
@@ -0,0 +1,234 @@
+}
+
+func NewUseCase(deps UseCaseDependencies) (*UseCase, error) {
+	if deps.WeatherService == nil {
```
I would place error checks somewhere else, maybe into separate method, idk because I would like to see use case creating only here

anastasiia-tokareva-hw replied:
`deps.Validate()`

YuliiaAndreieva commented on internal/core/subscription/entity.go:
```diff
@@ -0,0 +1,285 @@
+var emailRegex = regexp.MustCompile(`^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`)
+
+// NewToken creates a new token for a subscription
+func NewToken(subscriptionID uint, tokenType TokenType, expiresIn time.Duration) *Token {
```
I would separate it into folder token/entity

YuliiaAndreieva commented on internal/core/subscription/usecase.go:
```diff
@@ -0,0 +1,417 @@
+	return nil
+}
+
+func (uc *UseCase) Subscribe(ctx context.Context, params SubscribeParams) error {
```
long method is not readable, maybe separate it into some methods like 
```go
if existing != nil {
	return uc.handleExistingSubscription(ctx, existing, params)
}
return uc.createNewSubscription(ctx, params)

anastasiia-tokareva-hw replied:
Yes please

YuliiaAndreieva commented on internal/core/subscription/usecase.go:
```diff
@@ -0,0 +1,417 @@
+	confirmURL := fmt.Sprintf("%s/api/confirm/%s", baseURL, token)
+
+	return fmt.Sprintf(`
+		<h2>Confirm Your Weather Subscription</h2>
```
I think it's better not to know usecase what text will be in email, maybe create smth like EmailBuilder

anastasiia-tokareva-hw replied:
https://pkg.go.dev/text/template

YuliiaAndreieva commented on internal/core/weather/entity.go:
```diff
@@ -0,0 +1,92 @@
+	if strings.TrimSpace(w.Description) == "" {
+		return fmt.Errorf("description cannot be empty")
+	}
+	if w.Temperature < -273.15 {
```
make it const

YuliiaAndreieva commented on internal/ports/email_provider.go:
```diff
@@ -0,0 +1,16 @@
+	To      string
+	Subject string
+	Body    string
+	IsHTML  bool
```
Instead of using a boolean field (IsHTML) in the EmailParams struct, consider introducing a format field (e.g., an enum or set of constants) to represent the email format
```go
const (
	FormatText     EmailFormat = "text"
	FormatHTML     EmailFormat = "html"
	FormatMarkdown EmailFormat = "markdown"
)

type EmailParams struct {
	To      string
	Subject string
	Body    string
	Format  EmailFormat
}

YuliiaAndreieva commented on internal/ports/notification_service.go:
```diff
@@ -0,0 +1,27 @@
+// NotificationService defines the contract for notification operations
+type NotificationService interface {
+	SendWeatherUpdates(ctx context.Context, frequency string) error
+	CleanupExpiredTokens(ctx context.Context) error
```
sounds like responsibily for token service

Commit:
Implement caching.

Commit:
Refactors and fixes subscription and weather tests

Consolidates token validation error handling in subscription tests into a single, parameterized test function to improve readability and reduce code duplication.

Improves weather handler tests by introducing parameterized tests for validation errors.

Changes the default weather description to "N/A" when no weather data is available.

Remove architecture test passed comment on PR.

Commit:
Create validation middleware.

Commit:
Adds subscription filtering functionality

Implements a generic `Find` method in the subscription repository to allow filtering subscriptions based on various criteria such as ID, email, city, frequency, and confirmation status.
This change replaces the specific `GetConfirmedByFrequency`, `FindActiveByFrequency`, and `FindConfirmedByFrequency` methods with the new `Find` method, providing a more flexible and maintainable way to query subscriptions.
Updates the subscription service and use case to utilize the new filtering mechanism.
Adds comprehensive tests for the new filtering functionality, including validation of filter parameters.

Commit:
fix(token): decouples token creation from the repository

Introduces a TokenGenerator interface and a UUID-based implementation.
This allows for easier testing and more flexible token generation strategies.

Removes the token creation logic from the TokenRepository and instead uses the TokenGenerator to generate tokens, which are then saved by the repository.

Updates the subscription use case to use the new TokenGenerator for creating tokens.
Updates tests accordingly to mock the TokenGenerator.

anastasiia-tokareva-hw commented on internal/adapters/api/error_handler.go:
```diff
@@ -0,0 +1,133 @@
+	var apiErr *APIError
+	if errors.As(err, &apiErr) {
+		switch apiErr.Type {
+		case "API_VALIDATION_ERROR":
```
I would extract it to a const

anastasiia-tokareva-hw commented on internal/adapters/api/error_handler.go:
```diff
@@ -0,0 +1,133 @@
+}
+
+// getMetrics handles GET /api/metrics requests
+func (s *HTTPServerAdapter) getMetrics(c *gin.Context) {
```
why do you have `getMetrics` in `error_handler.go`?

anastasiia-tokareva-hw replied:
To me, it looks like parsing errors, error responses, etc. should be in one package (most likely common for other handlers) and handlers in a different package

anastasiia-tokareva-hw commented on internal/adapters/api/errors.go:
```diff
@@ -0,0 +1,29 @@
+// NewValidationError creates a new validation error for the API layer
+func NewValidationError(message string) *APIError {
+	return &APIError{
+		Type:    "API_VALIDATION_ERROR",
```
I see it for the second time, so it should definitely be a const 😁

anastasiia-tokareva-hw commented on internal/adapters/api/weather_handler.go:
```diff
@@ -0,0 +1,62 @@
+		City:        weatherData.City,
+	}
+
+	s.metricsCollector.IncrementCounter("api_responses_total", map[string]string{
```
I think general counters (happening on every request) should be part of middlewares

anastasiia-tokareva-hw commented on internal/adapters/database/subscription_repo.go:
```diff
@@ -0,0 +1,260 @@
+)
+
+// Error message constants for subscription repository
+const (
```
Ideally, you could have these in a separate file of this package
In general, don't be afraid to split the logic into different files

anastasiia-tokareva-hw commented on internal/adapters/database/subscription_repo.go:
```diff
@@ -0,0 +1,260 @@
+	var model SubscriptionModel
+	result := r.db.WithContext(ctx).Where("email = ? AND city = ?", email, city).First(&model)
+	if result.Error != nil {
+		if result.Error == gorm.ErrRecordNotFound {
```
I wonder if we can use `errors.Is` or `errors.As` instead of `==`

anastasiia-tokareva-hw commented on internal/adapters/database/subscription_repo.go:
```diff
@@ -0,0 +1,260 @@
+// Find retrieves subscriptions based on the provided filter criteria
+func (r *SubscriptionRepositoryAdapter) Find(ctx context.Context, filter ports.SubscriptionFilter) ([]*ports.SubscriptionData, error) {
+	// Validate that at least one filter criterion is provided
+	if filter.ID == nil && filter.Email == nil && filter.City == nil && filter.Frequency == nil && filter.Confirmed == nil {
```
should be a separate method `filter.Validate()`

anastasiia-tokareva-hw commented on internal/adapters/database/subscription_repo.go:
```diff
@@ -0,0 +1,260 @@
+
+	if filter.ID != nil {
+		if *filter.ID == 0 {
+			return nil, infrastructure.NewDatabaseError(ErrSubscriptionIDZero, nil)
```
move to validation?

anastasiia-tokareva-hw commented on internal/adapters/database/subscription_repo.go:
```diff
@@ -0,0 +1,260 @@
+	}
+	if filter.Email != nil {
+		if *filter.Email == "" {
+			return nil, infrastructure.NewDatabaseError(ErrEmailEmpty, nil)
```
part of `filter.Validate()`?

anastasiia-tokareva-hw commented on internal/adapters/external/accuweather_provider.go:
```diff
@@ -0,0 +1,66 @@
+		return nil, infrastructure.NewValidationError("city cannot be empty")
+	}
+
+	if p.apiKey == "" {
```
This check should be in `NewAccuWeatherProviderAdapter` not here

anastasiia-tokareva-hw commented on internal/adapters/external/cache_provider.go:
```diff
@@ -0,0 +1,211 @@
+		"providers_available": 1,
+		"primary_provider":    "weatherapi",
+		"status":              "active",
+		"cache_enabled":       true, // Add cache_enabled field
```
do you need this comment?

anastasiia-tokareva-hw commented on internal/adapters/external/cache_provider.go:
```diff
@@ -0,0 +1,211 @@
+// GetProviderInfo returns provider information
+func (m *WeatherMetricsAdapter) GetProviderInfo() map[string]interface{} {
+	// Get provider information from the provider manager
+	providerInfo := m.providerManager.GetProviderInfo()
```
this comment is a great example of code that doesn't need comments - you can infer all this info from the code itself

anastasiia-tokareva-hw commented on internal/adapters/external/email_provider.go:
```diff
@@ -0,0 +1,144 @@
+
+// SendEmail sends an email using SMTP with flexible authentication and TLS
+func (p *SMTPEmailProviderAdapter) SendEmail(ctx context.Context, params ports.EmailParams) error {
+	if params.To == "" {
```
`params.Validate()`

anastasiia-tokareva-hw commented on internal/adapters/external/email_provider.go:
```diff
@@ -0,0 +1,144 @@
+	msg := p.buildMessage(from, params.To, params.Subject, params.Body, params.IsHTML)
+	addr := fmt.Sprintf("%s:%d", p.host, p.port)
+
+	client, err := smtp.Dial(addr)
```
Could you possibly create `addr` and `client` in `NewSMTPEmailProviderAdapter` and reuse it here?

anastasiia-tokareva-hw commented on internal/adapters/external/email_provider.go:
```diff
@@ -0,0 +1,144 @@
+		}
+	}()
+
+	if ok, _ := client.Extension("STARTTLS"); ok {
```
encapsulate into `p.StartTLS`?

anastasiia-tokareva-hw commented on internal/adapters/external/redis_cache_provider.go:
```diff
@@ -0,0 +1,269 @@
+		WriteTimeout: time.Duration(config.WriteTimeout) * time.Second,
+	})
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
generally, it's a good idea to pass context from one place instead of creating it inside of the entities

anastasiia-tokareva-hw commented on internal/adapters/external/redis_cache_provider.go:
```diff
@@ -0,0 +1,269 @@
+}
+
+// RecordHit increments the cache hit counter
+func (r *RedisCacheProviderAdapter) RecordHit() {
```
I wonder why you need to call private methods from public methods without any added logic?

anastasiia-tokareva-hw commented on internal/adapters/external/weather_provider_manager.go:
```diff
@@ -0,0 +1,168 @@
+
+	// WeatherAPI provider
+	if config.WeatherAPIKey != "" {
+		providers["weatherapi"] = NewWeatherAPIProviderAdapter(WeatherAPIProviderParams{
```
to avoid setting string keys (basically, hardcoded info), you could've defined possible values: 

```go
type Provider string

const (
    WeatherAPI Provider = "weatherapi"
    OpenWeatherMap Provider = "openweathermap"
)

```


And then use it like: 

```go
	if config.WeatherAPIKey != "" {
		providers[WeatherAPI] = NewWeatherAPIProviderAdapter(WeatherAPIProviderParams{
```

or if you're keeping the string type as key, `string(WeatherAPI)`

anastasiia-tokareva-hw commented on internal/adapters/external/weather_provider_manager.go:
```diff
@@ -0,0 +1,168 @@
+}
+
+// GetProviderInfo returns information about configured providers
+func (m *WeatherProviderManagerAdapter) GetProviderInfo() map[string]interface{} {
```
I would suggest using specific types and not interfaces

You could've defined a struct with the needed fields and then referenced it here instead

anastasiia-tokareva-hw commented on internal/adapters/external/weather_providers.go:
```diff
@@ -0,0 +1,99 @@
+		}
+	}()
+
+	if resp.StatusCode != http.StatusOK {
```
I wonder if you could extract code that is the same for openweathermap and weather api and reuse it

anastasiia-tokareva-hw commented on internal/adapters/infrastructure/logger.go:
```diff
@@ -0,0 +1,46 @@
+
+// Info logs an info message
+func (l *SlogLoggerAdapter) Info(msg string, fields ...ports.Field) {
+	args := make([]interface{}, 0, len(fields)*2)
```
same code in every message - maybe it could be extracted to `extractArgs` method or something?

anastasiia-tokareva-hw commented on internal/adapters/infrastructure/metrics_collector.go:
```diff
@@ -0,0 +1,165 @@
+	}
+
+	if cacheStats, err := m.weatherMetrics.GetCacheMetrics(); err == nil {
+		cacheMetrics := map[string]interface{}{
```
should be a struct

anastasiia-tokareva-hw commented on internal/core/notification/usecase.go:
```diff
@@ -0,0 +1,234 @@
+	temperatureUnit := "°C"
+	humidityUnit := "%"
+
+	emailBody := fmt.Sprintf(`
```
check out template builders 

https://pkg.go.dev/text/template

anastasiia-tokareva-hw commented on internal/core/shared/errors.go:
```diff
@@ -0,0 +1,131 @@
+	}
+
+	// Check if it implements the NotFoundError interface
+	if nfe, ok := err.(interface{ IsNotFound() bool }); ok {
```
maybe 

```go
type NotFoundChecker interface {
	IsNotFound() bool
}

.......

if nfe, ok := err.(NotFoundChecker); ok {
	return nfe.IsNotFound()
}
```

anastasiia-tokareva-hw commented on internal/core/subscription/entity.go:
```diff
@@ -0,0 +1,285 @@
+// FromString converts string to Frequency enum
+func FrequencyFromString(s string) Frequency {
+	switch s {
+	case "hourly":
```
reused, so should be moved to const

anastasiia-tokareva-hw commented on internal/core/subscription/entity.go:
```diff
@@ -0,0 +1,285 @@
+func (t TokenType) String() string {
+	switch t {
+	case TokenTypeConfirmation:
+		return "confirmation"
```
please move to const

anastasiia-tokareva-hw commented on internal/core/subscription/entity.go:
```diff
@@ -0,0 +1,285 @@
+type TokenType int
+
+const (
+	TokenTypeUnknown TokenType = iota
```
You are mapping it to string later, why are you using enums? Any specific reasons?

anastasiia-tokareva-hw commented on internal/core/subscription/entity.go:
```diff
@@ -0,0 +1,285 @@
+	if s.Confirmed {
+		return false
+	}
+	return time.Since(s.CreatedAt) > 24*time.Hour
```
`const expiration = 24*time.Hour`

anastasiia-tokareva-hw commented on internal/core/subscription/entity.go:
```diff
@@ -0,0 +1,285 @@
+}
+
+func (sr *SubscriptionRequest) validateEmail() error {
+	if strings.TrimSpace(sr.Email) == "" {
```
looks like you could have a separate method `isEmptyString` that would trim string and compare it to an empty one

anastasiia-tokareva-hw commented on internal/core/subscription/usecase.go:
```diff
@@ -0,0 +1,456 @@
+}
+
+func NewUseCase(deps UseCaseDependencies) (*UseCase, error) {
+	if deps.SubscriptionRepo == nil {
```
`deps.Validate()`?

anastasiia-tokareva-hw commented on internal/core/subscription/usecase.go:
```diff
@@ -0,0 +1,456 @@
+		}
+
+		// If subscription exists but is not confirmed, update it with new parameters
+		if !subscription.IsExpired() {
```
better to have an early return if subscription is expired here

and then have the rest of the logic to avoid this huge block nesting

anastasiia-tokareva-hw commented on internal/core/subscription/usecase.go:
```diff
@@ -0,0 +1,456 @@
+		return fmt.Errorf("find token: %w", err)
+	}
+
+	if time.Now().After(tokenData.ExpiresAt) {
```
Maybe it could be `tokenData.IsExpired()` method

anastasiia-tokareva-hw commented on internal/core/subscription/usecase.go:
```diff
@@ -0,0 +1,456 @@
+	confirmToken, err := uc.createToken(ctx, CreateTokenParams{
+		SubscriptionID: subscription.ID,
+		TokenType:      TokenTypeConfirmation.String(),
+		ExpiresIn:      24 * time.Hour,
```
value should be a const

anastasiia-tokareva-hw commented on internal/core/weather/entity.go:
```diff
@@ -0,0 +1,92 @@
+
+// IsValid validates weather data
+func (w *Weather) IsValid() error {
+	if strings.TrimSpace(w.City) == "" {
```
could be a util method again

anastasiia-tokareva-hw commented on pkg/validation/validation.go:
```diff
@@ -0,0 +1,29 @@
+}
+
+// IsNotEmpty checks if string is not empty after trimming
+func IsNotEmpty(s string) bool {
```
You have this method! Why are you not using it in so many places?

anastasiia-tokareva-hw commented on pkg/validation/validation.go:
```diff
@@ -0,0 +1,29 @@
+
+// IsValidFrequency validates subscription frequency
+func IsValidFrequency(frequency string) bool {
+	return frequency == "hourly" || frequency == "daily"
```
If one time you decide to introduce `weekly` frequency, you will need to change code in a bunch of different files since you are hardcoding strings

anastasiia-tokareva-hw commented:
Good job! 
Your code has become much-much cleaner and more readable. 

Several times when I was writing comments, I would go to the next file and see that you are already doing what I want from you. So please make sure you are consistent in your approaches. If you are using best practices, use them everywhere. 

Once again, amazing job

anastasiia-tokareva-hw commented:
`map[string]interface{}` is an antipattern, please define structs

Commit:
fix(email_provider): make fixes according to review comments

Improves email sending reliability and configuration by:

- Adding configuration validation at startup to prevent runtime errors.
- Introducing a pre-computed server address for efficiency.
- Breaking down the SendEmail function into smaller, testable units for connection establishment, authentication, and message sending.
- Moving email parameter validation to the EmailParams struct.
- Adding comprehensive unit tests.

Commit:
refactor(memory_cache_provider): remove duplicate strings and create separate validation functions

Commit:
refactor(redis_cache): decouples cache implementation from metrics

Introduces a decorator pattern to separate caching logic from metrics collection.

This change enhances modularity and testability by allowing different
cache implementations to be used with or without metrics tracking..

Commit:
fix(errors): move get metrics out of error_handler

Commit:
refactor(adapters): move string definitions to separate file.

Commit:
refactor(weather): address PR review feedback

- Replace magic numbers with constants (DefaultHTTPTimeout)
- Move API key validation to constructors
- Use typed Provider constants instead of hardcoded strings
- Replace map[string]interface{} with structured ports.ProviderInfo
- Extract shared HTTPWeatherClient to eliminate code duplication
- Update mocks and tests for new interfaces

Commit:
refactor(subscription_service): remove duplicated function

Commit:
refactor(notification): move validation logic to separate function

Commit:
refactor(subscription): separate function into few ones for clarity

Commit:
refactor(weather_service): use interface for weather_service

Commit:
refactor(token): separate token to separate component

Commit:
refactor(email): create email builder

Commit:
refactor(weather_entity): move magic numbers to consts

Commit:
refactor(token): move token interface to separate service

Commit:
refactor: address PR review feedback

- Move API metrics to middleware for better separation
- Extract database constants to separate file
- Use errors.Is for proper error handling
- Centralize filter validation in Validate() method

Commit:
refactor(comments): remove redundant comments

Commit:
refactor(logger): move common functionality to a separate method

Commit:
refactor: implement PR feedback - consistent validation and early returns

- Add IsExpired() method to TokenData for better encapsulation
- Use validation.IsNotEmpty() consistently across all entities
- Replace hardcoded frequency strings with domain constants
- Refactor handleExistingSubscription to use early return pattern
- Remove unused IsValidFrequency to fix import cycle

Commit:
fix(arch_lint): fix to let tests pass

Commit:
feat(weather service): add weather service.

Commit:
feat(user service): add user service

Commit:
feat(subscription service): add subscription service

Commit:
feat(notification service): add notification service

Commit:
feat(infrastructure): add docker configuration

Commit:
feat(api gateway): add api gateway and restore WEB UI

Commit:
fix(config): remove dafault parameters

Commit:
refactor(subscription): use notification service to send emails

Commit:
fix(tests): missing mocks

Commit:
fix(integration tests)

Commit:
fix(e2e tests)

Commit:
feat(logger): add structured logging

Commit:
refactor(logging): replace old adapter with new one

Commit:
refactor(readme): update readme with metrics configuration and policies

Commit:
fix(unit test)

Commit:
refactor(metrics): extract helper methods for prometheus metric creation

Mishok7889 committed:
Merge pull request #13 from GenesisEducationKyiv/hw11-observability

Hw11 observability

Commit:
fix: add consistent error wrapping across all service layers

Commit:
refactor: remove debug comments from notification service config

Commit:
fix: replace silenced errors with proper error logging

Commit:
refactor: replace hardcoded 'healthy' status with constants

Commit:
fix: update HTTP status code boundaries from 300 to 400

Commit:
refactor: eliminate DRY violation in Headers() methods

Commit:
refactor: remove redundant constants anti-pattern

Commit:
refactor: move inline struct definitions to named types

Commit:
refactor: add IsExpired method to TokenModel

Commit:
refactor: optimize gRPC slice usage - use values instead of pointers

Commit:
refactor: use early returns to reduce nesting in Shutdown methods

Commit:
refactor: extract HTTP request preparation to reduce duplication

Commit:
refactor: replace interface{} with any for modern Go

Commit:
fix: resolve gRPC WeatherData pointer type mismatch

Commit:
refactor(e2e): change error label.

Mishok7889 committed:
Merge pull request #12 from GenesisEducationKyiv/hw8-9-microservices

Hw9-10 microservices
